crates/wasmtime/src/component/func/host.rs

```diff
@@ -330,3 +365,243 @@ macro_rules! impl_into_component_func {
+                        ty,
+                        &memory,
+                        &memory.as_slice()[ty.next_field(&mut offset)..]
+                            [..ty.size_and_alignment().size],
+                    )
+                })
+                .collect::<Result<Box<[_]>>>()?;
+            let ret = closure(cx.as_context_mut(), &args)?;
+            result.check(&ret)?;
+            let mut memory = MemoryMut::new(cx.as_context_mut(), &options);
+            let ptr = validate_inbounds_dynamic(
+                result.size_and_alignment(),
+                memory.as_slice_mut(),
+                ret_ptr,
+            )?;
+            flags.set_may_leave(false);
+            ret.store(&mut memory, ptr)?;
+        }
+    }
+
+    flags.set_may_leave(true);
+
+    return Ok(());
+}
+
+fn validate_inbounds_dynamic(
+    SizeAndAlignment { size, alignment }: SizeAndAlignment,
+    memory: &[u8],
+    ptr: &ValRaw,
+) -> Result<usize> {
+    // FIXME: needs memory64 support
+    let ptr = usize::try_from(ptr.get_u32())?;
+    if ptr % usize::try_from(alignment)? != 0 {
+        bail!("pointer not aligned");
+    }
+    let end = match ptr.checked_add(size) {
+        Some(n) => n,
+        None => bail!("pointer size overflow"),
+    };
+    if end > memory.len() {
+        bail!("pointer out of bounds")
+    }
+    Ok(ptr)
+}
+
+struct Types {
+    params: Box<[Type]>,
+    result: Type,
+}
+
+struct DynamicContext<F> {
+    func: F,
+    types: Arc<Mutex<Option<Types>>>,
+}
+
+// Implement for functions without a leading `StoreContextMut` parameter
+#[allow(non_snake_case)]
+impl<T, F> IntoComponentFunc<T, &[Val], Val> for F
+where
+    F: Fn(&[Val]) -> Result<Val> + Send + Sync + 'static,
+{
+    extern "C" fn entrypoint(
+        cx: *mut VMOpaqueContext,
+        data: *mut u8,
+        flags: InstanceFlags,
+        memory: *mut VMMemoryDefinition,
+        realloc: *mut VMCallerCheckedAnyfunc,
+        string_encoding: StringEncoding,
+        storage: *mut ValRaw,
+        storage_len: usize,
+    ) {
+        let data = data as *const DynamicContext<F>;
+        unsafe {
+            handle_result(|| {
+                call_host_dynamic::<T, _>(
+                    (*data).types.lock().unwrap().as_ref().unwrap(),
+                    cx,
+                    flags,
+                    memory,
+                    realloc,
+                    string_encoding,
+                    std::slice::from_raw_parts_mut(storage, storage_len),
+                    |_, values| ((*data).func)(values),
+                )
+            })
+        }
+    }
+
+    fn into_host_func(self) -> Arc<HostFunc> {
+        let entrypoint = <Self as IntoComponentFunc<T, &[Val], Val>>::entrypoint;
+        HostFunc::new_dynamic(self, entrypoint)
+    }
+}
+
+// Implement for functions with a leading `StoreContextMut` parameter
+#[allow(non_snake_case)]
+impl<T, F> IntoComponentFunc<T, (StoreContextMut<'_, T>, &[Val]), Val> for F
```

Comment:
Ah sorry I forgot to review this in the initial first pass. We won't want to have this implementation though because the `IntoComponentFunc` impl is only intended for host functions where the component-type of the function can be statically inferred. As you have probably surmised with the lock/etc in typechecking for these functions the "dynamic" interface here doesn't really match the interface for static functions.

Instead what we'll want to do is something along the lines of the difference between `wasmtime::Func::{wrap,new}` where here you're defining the `new` equivalent. That will grow a new function `component::Linker::func_new` which will take the type of the function as its argument as well as a closure. The return values of the closure are then type-checked against the original type at runtime before yielding the result back to the original component.

This may be worth splitting to its own PR if you'd prefer. I'm also happy to review it here as well though.

Reply:
Would you mind sketching out an example of how your proposed API would be used?  I'm specifically interested in where the user would get the type of the function to pass to `component::Linker::func_new`.  Would they query the `Component` for the things it imports, find the function of interest, and get the type from that?

Reply:
Yeah that is indeed a good question. In core wasm users would do `FuncType::new(...)` and create their own function type. In the fullness of time I'd expect the same operation to be available to components where function types could be manufactured in addition to being learned from components. That's a big missing piece of the embedding API for components right now though so, yeah, getting it from a component seems the most reasonable for now.

Reply:
@alexcrichton I've just pushed an update per your suggested API.  Let me know what you think.